File: library/src/test/java/de/westnordost/countryboundaries/CountryBoundariesCellTest.java
package de.westnordost.countryboundaries;

import org.junit.Test;

import java.util.Arrays;
import java.util.Collections;

import static org.junit.Assert.assertEquals;
import static org.junit.Assert.assertFalse;
import static org.junit.Assert.assertTrue;

public class CountryBoundariesCellTest
{

	private static CountryAreas[] B = new CountryAreas[]{ new CountryAreas("B",
			new Point[][]{new Point[]{p(0, 0), p(0, 2), p(2, 2), p(2, 0)}},
			new Point[][]{})
	};

	@Test public void getDefiniteIds()
	{
		assertEquals(
			Arrays.asList("A","C"),
			cell(new String[]{"A","C"}, null).getIds(0,0)
		);
	}

	@Test public void getInGeometryIds()
	{
		assertEquals(
			Arrays.asList("B"),
			cell(null, B).getIds(1,1)
		);
	}

	@Test public void dontGetOutOfGeometryIds()
	{
		assertEquals(
			Collections.emptyList(),
			cell(null, B).getIds(4,4)
		);
	}

	@Test public void getDefiniteAndInGeometryIds()
	{
		assertEquals(
			Arrays.asList("A","B"),
			cell(new String[]{"A"},B).getIds(1,1)
		);
	}

	@Test public void getAllIds()
	{
		assertEquals(
			Arrays.asList("A","B"),
			cell(new String[]{"A"},B).getAllIds()
		);
	}

	@Test public void isInAnyDefinitely()
	{
		assertTrue(cell(new String[]{"A"},null).isInAny(0,0,Arrays.asList("B","A")));
	}

	@Test public void isInAnyDefinitelyNot()
	{
		assertFalse(cell(new String[]{"A"},null).isInAny(0,0,Arrays.asList("B")));
	}

	@Test public void isInAnyInGeometry()
	{
		assertTrue(cell(null, B).isInAny(1,1,Arrays.asList("B")));
	}

	@Test public void isInAnyOutOfGeometry()
	{
		assertFalse(cell(null, B).isInAny(4,4,Arrays.asList("B")));
	}

	/* Helpers */

	private static CountryBoundariesCell cell(String[] containingIds, CountryAreas[] intersecting)
	{
		return new CountryBoundariesCell(
				containingIds == null ? Collections.emptyList() : Arrays.asList(containingIds),
				intersecting == null ? Collections.emptyList() : Arrays.asList(intersecting)
		);
	}

	private static Point p(double x, double y)
	{
		return new Point(Fixed1E7.doubleToFixed(x),Fixed1E7.doubleToFixed(y));
	}
}

File: library/src/commonTest/kotlin/de/westnordost/countryboundaries/CountryBoundariesCellTest.kt
package de.westnordost.countryboundaries

import kotlin.test.*

internal class CountryBoundariesCellTest {
    @Test fun definite_ids() {
		assertEquals(
			listOf("A", "C"),
			cell(containing = listOf("A", "C")).getIds(p(0, 0))
		)
	}

    @Test fun in_geometry_ids() {
		assertEquals(
			listOf("B"),
			cell(intersecting = listOf(B)).getIds(p(1, 1))
		)
	}

    @Test fun dont_get_out_of_geometry_ids() {
        assertEquals(
            emptyList(),
            cell(intersecting = listOf(B)).getIds(p(4, 4))
        )
    }

    @Test fun definite_and_in_geometry_ids() {
		assertEquals(
			listOf("A", "B"),
			cell(containing = listOf("A"), intersecting = listOf(B)).getIds(p(1, 1))
		)
	}

    @Test fun all_ids() {
		assertEquals(
			listOf("A", "B"),
			cell(containing = listOf("A"), intersecting = listOf(B)).getAllIds()
		)
	}

    @Test fun is_in_any_definitely() {
		assertTrue(
			cell(containing = listOf("A")).isInAny(p(0, 0), listOf("B", "A"))
		)
	}

    @Test fun is_in_any_definitely_not() {
		assertFalse(
			cell(containing = listOf("A")).isInAny(p(0, 0), listOf("B"))
		)
	}

    @Test fun is_in_any_in_geometry() {
		assertTrue(
			cell(intersecting = listOf(B)).isInAny(p(1, 1), listOf("B"))
		)
	}

    @Test fun is_in_any_out_of_geometry() {
		assertFalse(
			cell(intersecting = listOf(B)).isInAny(p(4, 4), listOf("B"))
		)
	}
}

private val B = CountryAreas(
	id = "B",
	outer = listOf(listOf(p(0, 0), p(0, 2), p(2, 2), p(2, 0))),
	inner = listOf()
)

private fun p(x: Int, y: Int) = Point(x.toUShort(), y.toUShort())

private fun cell(containing: List<String> = emptyList(), intersecting: List<CountryAreas> = emptyList()) =
    CountryBoundariesCell(containing, intersecting)
